### PR TITLE
hcm: forbid use of detection extensions with use_remote_addr/xff_num_trusted_hops

### LINKED ISSUE #17554: IP detection regresion
From https://groups.google.com/g/envoy-users/c/vGD0YT8aF2Q

Hi all,

tl;dr: The latest Envoy release contains a regression related to how IP detection works when xff_num_trusted_hops > 0 and use_remote_address is true. To avoid it, do not use the new original_ip_detection_extensions field. Instead, keep using the xff_num_trusted_hops deprecated field until the next release or the corresponding fix becomes available.

After support for IP detection extensions was added [0], the xff_num_trusted_hops [1] field was deprecated in favor of the equivalent extension. Using the XFF extension [2] along with the use_remote_address option [3] will break detection because the detection code still relies on the deprecated field as opposed to using the field in the extension itself. More specifically, the configured value in the extension for xff_num_trusted_hops – when it’s greater than zero – is ignored when used in conjunction with use_remote_address. This means that you will end up with the remote address as the final address as opposed to it being extracted from the x-forwarded-for header. To fix this we will retire the deprecation around the xff_num_trusted_hops field, as well as preventing it and use_remote_address from being used with extensions given that this yields undefined behavior. 

We’d like to thank Kateryna Nezdolii for reporting the issue and helping us debug it. 

Thanks,

The Envoy Team

[0] https://github.com/envoyproxy/envoy/pull/14855
[1] https://github.com/envoyproxy/envoy/blob/537cf1a8482fbdd74100e708547086ff443e6250/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L520
[2] https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/http/original_ip_detection/xff/v3/xff.proto 
[3] https://github.com/envoyproxy/envoy/blob/537cf1a8482fbdd74100e708547086ff443e6250/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L496

### PR DESCRIPTION
**Pull request by rgs1**
Mixing extensions with previously existing knobs leads to undefined behavior,
so this removes the deprecation around xff_num_trusted_hops and ensures that
it's not mixed with extensions.

Note that a unit test already exists for the original bug report, where
use_remote_address is used with xff_num_trusted_hops > 0. However it uses
the old knob instead of XFF extension. Given this is now forbidden,
there's no need for additional tests wrt that config combination.

Fixes #17554.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
One possible wrinkle is:

```
  // Edge request is the request from external clients to front Envoy.
  // Request from front Envoy to the internal service will be treated as not edge request.
  const bool edge_request = !internal_request && config.useRemoteAddress();
  ```
  
which means, if you are using an extension (or a list of them) you won't be able to treat requests as edge. However, if you are using an extension you most likely are not on the edge. So we can punt solving this one, until an extension writer comes along with the need to label requests as edge.

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -40,6 +40,7 @@ Bug Fixes
 
 * access log: fix `%UPSTREAM_CLUSTER%` when used in http upstream access logs. Previously, it was always logging as an unset value.
 * cluster: delete pools when they're idle to fix unbounded memory use when using PROXY protocol upstream with tcp_proxy. This behavior can be temporarily reverted by setting the ``envoy.reloadable_features.conn_pool_delete_when_idle`` runtime guard to false.
+* hcm: remove deprecation for xff_num_trusted_hops and forbid mixing ip detection extensions with old related knobs.
```
Comment (mattklein123) [inline docs/root/version_history/current.rst:43]: ref link to xff_num_trusted_hops field.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks this makes sense to me. I feel like there should be more documentation around this. Can we explain this somewhere in the arch overview or HTTP docs where we already talk about XFF, num trusted hops, etc.?

/wait

**Comment by rgs1** [comment]
> Thanks this makes sense to me. I feel like there should be more documentation around this. Can we explain this somewhere in the arch overview or HTTP docs where we already talk about XFF, num trusted hops, etc.?
> 

Added some, looking for additional spots that could use warnings around the mix of extensions and the pre-existing knobs.

**Review by mattklein123 — DISMISSED** [review]
Thanks!

**Comment by mattklein123** [comment]
Oops looks like you need a main merge.

/wait

**Comment by rgs1** [comment]
Err this seems unrelated:

```
[0 / 1] [Prepa] BazelWorkspaceStatusAction stable-status.txt
Target //tools/testing:python_coverage up-to-date:
  bazel-bin/tools/testing/python_coverage
INFO: Elapsed time: 0.275s, Critical Path: 0.11s
INFO: 3 processes: 3 internal.
INFO: Build completed successfully, 3 total actions
INFO: Running command line: bazel-bin/tools/testing/python_coverage /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/tools/testing/all_pytests.runfiles/envoy/.coverage-envoy /source/generated/tooling
INFO: Build completed successfully, 3 total actions
PytestChecker ERROR Summary
--------------------------------------------------------------------------------
PytestChecker pytests:
//tools/extensions:pytest_extensions_check failed

PytestChecker ERROR {'success': 17, 'errors': 1, 'warnings': 0, 'failed': {'pytests': 1}, 'warned': {}, 'succeeded': {'pytests': 17}}
```

cc: @phlax

**Comment by rgs1** [comment]
In the case of ARM:

```
//test/integration:sds_dynamic_integration_test                          FAILED in 2 out of 2 in 15.0s
  Stats over 2 runs: max = 15.0s, min = 14.8s, avg = 14.9s, dev = 0.1s
  /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/aarch64-opt/testlogs/test/integration/sds_dynamic_integration_test/test.log
  /build/tmp/_bazel_envoybuild/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/aarch64-opt/testlogs/test/integration/sds_dynamic_integration_test/test_attempts/attempt_1.log
  ```

**Review by adisuissa — COMMENTED** [review]
/lgtm api

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**